### PR TITLE
Amesos2 : add interface to MKL cluster sparse solver

### PR DESCRIPTION
**Pull request by iyamazaki**
@trilinos/amesos2

## Motivation

This PR adds an interface to MKL's Cluster Sparse Solver.

## Additional Information

Cluster Sparse Solver is enabled with Pardiso_MKL and MPI, e.g.,
```
    -D TPL_ENABLE_MPI:BOOL=ON \
    \
    -D TPL_ENABLE_METIS:BOOL=ON \
    -D   METIS_INCLUDE_DIRS="${METIS_ROOT}/include" \
    -D   METIS_LIBRARY_DIRS="${METIS_ROOT}/lib64" \
    -D TPL_ENABLE_MKL:BOOL=ON \
    -D   TPL_MKL_LIBRARIES="-qmkl=parallel;-static-intel" \
    -D   TPL_MKL_INCLUDE_DIRS="${MKLROOT}/include" \
    -D TPL_ENABLE_PARDISO_MKL:BOOL=ON \
    -D   PARDISO_MKL_LIBRARY_DIRS="${MKLROOT}/lib/intel64" \
    -D   PARDISO_MKL_INCLUDE_DIRS="${MKLROOT}/include" \
    -D TPL_ENABLE_BLAS:BOOL=ON \
    -D   BLAS_LIBRARY_DIRS="${MKLROOT}/lib/intel64" \
    -D   BLAS_LIBRARY_NAMES="mkl_intel_lp64;mkl_intel_thread;mkl_core;mkl_blacs_intelmpi_lp64" \
    -D TPL_ENABLE_LAPACK:BOOL=ON \
    -D   LAPACK_LIBRARY_DIRS="${MKLROOT}/lib/intel64" \
    -D   LAPACK_LIBRARY_NAMES="mkl_intel_lp64;mkl_intel_thread;mkl_core;mkl_blacs_intelmpi_lp64" \
```

### CODE REVIEW

File: packages/amesos2/src/Amesos2_CssMKL.cpp [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+
+
+#include "Amesos2_CssMKL_decl.hpp"
+
+#ifdef HAVE_AMESOS2_EXPLICIT_INSTANTIATION
```
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_CssMKL.cpp:47]: There is much more simplified logic for handling all this, for example with Amesos2_KLU2.cpp. Would be good to change to simplify and reduce likelihood of copy-paste errors, but I'll let you decide if that is better as a second round PR

File: packages/amesos2/src/Amesos2_CssMKL_FunctionMap.hpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+   \file   Amesos2_NewSolver_FunctionMap.hpp
+   \author Eric Bavier <etbavie@sandia.gov>
+   \date   Wed Jul 27 12:53:10 MDT 2011
+
+   \brief  Template for providing a mechanism to map function calls to the
+           correct Solver function based on the scalar type of Matrices and
+           MultiVectors
+*/
+
+#ifndef AMESOS2_NEWSOLVER_FUNCTIONMAP_HPP
```
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_CssMKL_FunctionMap.hpp:55]: This needs to be redefined (L55-56) to match the header file name, hopefully there isn't a collision with another Amesos2 file having the same non-updated name

File: packages/amesos2/src/Amesos2_CssMKL_FunctionMap.hpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			 const MPI_Fint  * comm  , PMKL::_INTEGER_t* error)
+    {
+      PMKL::cluster_sparse_solver(pt, maxfct, mnum, mtype, phase, n, a, ia, ja,
+		                  perm, nrhs, iparm, msglvl, b, x, comm, error);
+    }
+  };
+
+
+  template <>
+  struct FunctionMap<CssMKL,long long int>
```
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_CssMKL_FunctionMap.hpp:106]: Reading the comment above (line ~83), does `long long int` match to `pardiso_64`?

File: packages/amesos2/src/Amesos2_CssMKL_def.hpp [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+  {
+    int_t error = 0;
+    {
+#ifdef HAVE_AMESOS2_TIMERS
+      Teuchos::TimeMonitor symbFactTimer( this->timers_.symFactTime_ );
+#endif
+
+      int_t phase = 11; // Analysis
+      void *bdummy, *xdummy;
+      const MPI_Fint CssComm = CssComm_;
```
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_CssMKL_def.hpp:170]: Neat, I was not aware of `MPI_Fint` until seeing this

**Review by ndellingwood — CHANGES_REQUESTED** [review]
I didn't dive fully into the details (I'm not familiar with pardiso solver), but the the Amesos2 wrappers look reasonable; only change needed is to fix the header guards in one of the files
@iyamazaki can you also add a configuration you used to build + test this for reproducibility? Thanks!

File: packages/amesos2/src/Amesos2_CssMKL.cpp (hunk 1 above)
Comment (iyamazaki) [reply packages/amesos2/src/Amesos2_CssMKL.cpp:47]: The simpler version lead to build errors like this:
```
Amesos2_MatrixAdapter_def.hpp(195): error: class "Amesos2::ConcreteMatrixAdapter<KokkosSparse::CrsMatrix<double, int, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>, void, size_t={unsigned long}>>" has no member "getLocalNNZ_impl"
      return static_cast<const adapter_t*>(this)->getLocalNNZ_impl();
```

File: packages/amesos2/src/Amesos2_CssMKL_FunctionMap.hpp (hunk 3 above)
Comment (iyamazaki) [reply packages/amesos2/src/Amesos2_CssMKL_FunctionMap.hpp:106]: It looks like it maps to `INTEGER*8`.

**Review by ndellingwood — APPROVED** [review]
Thanks @iyamazaki !

**Pull request merged**